### DISCUSSION TITLE
Shift X to reduce selection to the line above

### DISCUSSION THREAD

**Discussion by tqwewe** [helix-editor/helix#6943] [Ideas] [answered]
Helix currently allows users to select multiple adjacent lines using the `x` key. However, there might be instances where users accidentally select one extra line and wish to deselect the current line and navigate back up.

A potential solution is to introduce the `Shift X` key binding for this purpose, along with a new command, possibly named `reduce_line_above`. It is worth noting that `Shift X` is currently assigned to "Extend selection to line bounds (line-wise selection)". However, the functionality and difference between `x` and `Shift X` is not clear to me.

**Comment by CptPotato** [comment]
> However, the functionality and difference between `x` and `Shift X` is not clear to me.

<kbd>Shift</kbd> <kbd>X</kbd> will not extend the selection to multiple lines.

So when you have a cursor on an empty line the behavior between <kbd>X</kbd> and <kbd>Shift</kbd> <kbd>X</kbd> is different (<kbd>X</kbd> will extend to the next line in that case).

**Reply by tqwewe** [reply]
Thank you for clarifying.
I still would love a way to reduce the selection up a line

**Reply by tqwewe** [reply]
What does <kbd>Alt</kbd> <kbd>x</kbd> do?

I've dug into the source code and played around with it, but I honestly can't get my head around what it does exactly, and if anyone would actually use it.

**Reply by jnns** [reply]
`Alt-x` (shrink_to_line_bounds) removes those lines from the selection that don't cover whole lines.

<img width="71" alt="image" src="https://github.com/user-attachments/assets/68d22a35-9f96-47d2-b5f3-e423eeaac63a" />
<img width="71" alt="image" src="https://github.com/user-attachments/assets/860eac7c-3bf6-4674-9981-8008047ebaff" />

**Comment by jesusmgg** [comment]
Try this: press v and then k. It will deselect the last line.

Then v again to exit select mode if you want.

**Reply by tqwewe** [reply]
This almost works, but it doesn't extend to the end of the line after pressing k.
For the behavior I'm describing, you'd have to do `v` `k` `X` `v`.

**Reply by gabydd** [reply]
You can create a custom bind to ``["extend_line_up", "extend_to_line_bounds"]`` which should give you the functionality you want

**Comment by tqwewe** [comment, marked as answer]
There seems to already be a command for extending a line up.

All I had to do was add this to my config:
```toml
[keys.normal]
X = ["extend_line_up", "extend_to_line_bounds"]
A-x = "extend_to_line_bounds"

[keys.select]
X = ["extend_line_up", "extend_to_line_bounds"]
A-x = "extend_to_line_bounds"
```
The <kbd>Alt</kbd> <kbd>x</kbd> keybindings are there so I still have access to the old <kbd>Shift</kbd> <kbd>x</kbd> command.
Though this removes the shrink_line_bounds bindings, which I honestly have never understood what its for and how it works.

**Reply by gabydd** [reply]
Visual line in this case actually means it moves by softwrapped lines that visually look like different lines, the commands that start with extend are the actual visual mode commands

**Reply by tqwewe** [reply]
Oh so both these keys should be mapped to `extend_line_up`?

**Reply by michaeljones** [reply]
Thanks for sharing. I've often been annoyed by going one over with `x` presses and would be keen to have something to address it easily. 

I can't see `extend_line_up` in the docs? Even the `master` version of the docs? Am I missing something? Or is it missing?

**Reply by tqwewe** [reply]
@michaeljones I only came across it in the source code. There's a link to it at the bottom of <https://docs.helix-editor.com/remapping.html>:
https://github.com/helix-editor/helix/blob/master/helix-term/src/commands.rs

I just opened that file and did a ctrl find for `"up"`

**Reply by KaaiH** [reply]
I think the proper way to get X to expand the section upwards would be 
`X = "extend_line_above"`.

**Comment by gabydd** [comment]
Yep, plus you should add ``extend_to_line_bounds`` for when the lower line is smaller then the line above

**Comment by matuusu** [comment]
If you selected one or few extra lines and wish to deselect the extra lines, just enter visual mode `v` and use `k` or `up` to shrink selection.

**Comment by chtenb** [comment]
Using the new `select_line_below` and `select_line_above` commands this becomes easier. For reference https://github.com/helix-editor/helix/pull/9080#issuecomment-2008964831
